duplicate names in source file or source directory break pgwatch

**Describe the bug**
While trying to play around with multiple sources defined in a directory, I ran into a problem.
At first I thought it was a problem with pgwatch being unable to work with the directory, but it turns out pgwatch breaks when a source name is reused.
This can even happen within a single source.yaml file.

**To Reproduce**
Steps to reproduce the behavior:
1.Create a simple source.yaml config file with two instances but "forget" to give a new unique name:
```
- name: test
  kind: postgres
  connstr: postgresql:///postgres?host=/var/run/postgresql
  is_superuser: true
  preset_metrics: exhaustive
  is_enabled: true
- name: test
  kind: postgres
  connstr: postgresql:///pgbench?host=/var/run/postgresql
  is_superuser: true
  preset_metrics: exhaustive
  is_enabled: true
```
2. Start pgwatch
```
pgwatch --sources test.yaml --sink postgresql:///postgres?host=/var/run/postgresql
```
 
4. See error
```
2025-01-31 18:34:37.975 [INFO] [sink:postgresql:///postgres?host=/var/run/postgresql] Initialising measurements database...
2025-01-31 18:34:37.980 [INFO] [sink:postgresql:///postgres?host=/var/run/postgresql] Measurements sink activated
2025-01-31 18:34:37.981 [INFO] [metrics:75] [presets:17] [sources:2] sources and metrics refreshed
2025-01-31 18:34:37.986 [ERROR] [callstack:goroutine 1 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.23.4/x64/src/runtime/debug/stack.go:26 +0x5e
main.main.func1()
	/home/runner/work/pgwatch/pgwatch/cmd/pgwatch/main.go:49 +0xb2
panic({0xe241a0?, 0x1e0a6c0?})
	/opt/hostedtoolcache/go/1.23.4/x64/src/runtime/panic.go:785 +0x132
github.com/cybertec-postgresql/pgwatch/v3/internal/reaper.GetMonitoredDatabaseSettings({0x16abd10, 0xc000399080}, {0xc000012a34, 0x4}, {0xc000012a48, 0x8}, 0x1)
	/home/runner/work/pgwatch/pgwatch/internal/reaper/database.go:237 +0x689
github.com/cybertec-postgresql/pgwatch/v3/internal/reaper.(*Reaper).Reap(0xc000399350, {0x16abd10, 0xc000399080})
	/home/runner/work/pgwatch/pgwatch/internal/reaper/reaper.go:123 +0x144e
main.main()
	/home/runner/work/pgwatch/pgwatch/cmd/pgwatch/main.go:100 +0x8eb
] runtime error: invalid memory address or nil pointer dereference
```

**Expected behavior**
pgwatch should exit and tell me that I made a mistake of using the same instance name more than once.


debug log below:
```
postgres@pgwatch-debian:~$ pgwatch --sources test.yaml --sink postgresql:///postgres?host=/var/run/postgresql --log-level=debug
2025-01-31 18:40:08.789 [DEBUG] opts: &{Sources:{Sources:test.yaml Refresh:120 Groups:[] MinDbSizeMB:0 MaxParallelConnectionsPerDb:4 TryCreateListedExtsIfMissing:} Metrics:{Metrics: CreateHelpers:false DirectOSStats:false InstanceLevelCacheMaxSeconds:30 EmergencyPauseTriggerfile:/tmp/pgwatch-emergency-pause} Sinks:{Sinks:[postgresql:///postgres?host=/var/run/postgresql] BatchingDelay:250ms Retention:14 RealDbnameField:real_dbname SystemIdentifierField:sys_id} Logging:{LogLevel:debug LogFile: LogFileFormat:json LogFileRotate:false LogFileSize:100 LogFileAge:0 LogFileNumber:0} WebUI:{WebDisable:false WebAddr::8080 WebUser: WebPassword:} Help:false SourcesReaderWriter:<nil> MetricsReaderWriter:<nil> ExitCode:0 CommandCompleted:false} (/home/runner/work/pgwatch/pgwatch/cmd/pgwatch/main.go:76 main.main)
2025-01-31 18:40:08.803 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [database:postgres] [host:/var/run/postgresql] [pid:1192] [port:5432] [time:4.273858ms] Connect
2025-01-31 18:40:08.804 [INFO] [sink:postgresql:///postgres?host=/var/run/postgresql] Initialising measurements database... (sinks/postgres.go:43 sinks.NewWriterFromPostgresConn.func1)
2025-01-31 18:40:08.805 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:SELECT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname = $1)] [alreadyPrepared:false] [name:stmtcache_f6e8c26f6f8a335150a8a635aaf7985e6f591c0ddc59743d] [pid:1192] [time:829.372µs] Prepare
2025-01-31 18:40:08.806 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:SELECT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname = $1)] [args:[admin]] [commandTag:SELECT 1] [pid:1192] [time:2.420774ms] Query
2025-01-31 18:40:08.807 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:SELECT schema_type = 'timescale' FROM admin.storage_schema_type] [alreadyPrepared:false] [name:stmtcache_990ec515fab3f2cb3f7a4f0055b689d89ebc5c34fd994edf] [pid:1192] [time:957.289µs] Prepare
2025-01-31 18:40:08.808 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:SELECT schema_type = 'timescale' FROM admin.storage_schema_type] [args:[]] [commandTag:SELECT 1] [pid:1192] [time:2.016756ms] Query
2025-01-31 18:40:08.809 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [alreadyPrepared:false] [name:stmtcache_ecdc2d3d7f2a3688badd01ad8e4adba4b0cacc4707d393a8] [pid:1192] [time:417.12µs] Prepare
2025-01-31 18:40:08.812 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [args:[sproc_changes]] [commandTag:SELECT 1] [pid:1192] [time:3.575236ms] Query
2025-01-31 18:40:08.813 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [args:[table_changes]] [commandTag:SELECT 1] [pid:1192] [time:819.945µs] Query
2025-01-31 18:40:08.814 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [args:[index_changes]] [commandTag:SELECT 1] [pid:1192] [time:594.258µs] Query
2025-01-31 18:40:08.815 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [args:[privilege_changes]] [commandTag:SELECT 1] [pid:1192] [time:796.912µs] Query
2025-01-31 18:40:08.817 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [args:[object_changes]] [commandTag:SELECT 1] [pid:1192] [time:1.156226ms] Query
2025-01-31 18:40:08.818 [DEBUG] [sink:postgresql:///postgres?host=/var/run/postgresql] [sql:select admin.ensure_dummy_metrics_table($1)] [args:[configuration_changes]] [commandTag:SELECT 1] [pid:1192] [time:560.496µs] Query
2025-01-31 18:40:08.818 [INFO] [sink:postgresql:///postgres?host=/var/run/postgresql] Measurements sink activated (sinks/multiwriter.go:54 sinks.NewMultiWriter)
2025-01-31 18:40:08.818 [INFO] [metrics:75] [presets:17] [sources:2] sources and metrics refreshed (reaper/reaper.go:95 reaper.(*Reaper).Reap)
2025-01-31 18:40:08.818 [DEBUG] [source:test] [metric:map[]] [config:{ []          []}] [tags:map[]]  (reaper/reaper.go:108 reaper.(*Reaper).Reap)
2025-01-31 18:40:08.822 [DEBUG] [database:] [host:/var/run/postgresql] [pid:1193] [port:5432] [time:3.777369ms] Connect
2025-01-31 18:40:08.823 [DEBUG] [source:test] [kind:postgres] determining DB version and recovery status... (reaper/database.go:201 reaper.GetMonitoredDatabaseSettings)
2025-01-31 18:40:08.824 [ERROR] [callstack:goroutine 1 [running]:
runtime/debug.Stack()
	/opt/hostedtoolcache/go/1.23.4/x64/src/runtime/debug/stack.go:26 +0x5e
main.main.func1()
	/home/runner/work/pgwatch/pgwatch/cmd/pgwatch/main.go:49 +0xb2
panic({0xe241a0?, 0x1e0a6c0?})
	/opt/hostedtoolcache/go/1.23.4/x64/src/runtime/panic.go:785 +0x132
github.com/cybertec-postgresql/pgwatch/v3/internal/reaper.GetMonitoredDatabaseSettings({0x16abd10, 0xc0003598c0}, {0xc0004733cc, 0x4}, {0xc0004733e0, 0x8}, 0x1)
	/home/runner/work/pgwatch/pgwatch/internal/reaper/database.go:237 +0x689
github.com/cybertec-postgresql/pgwatch/v3/internal/reaper.(*Reaper).Reap(0xc000359d10, {0x16abd10, 0xc0003598c0})
	/home/runner/work/pgwatch/pgwatch/internal/reaper/reaper.go:123 +0x144e
main.main()
	/home/runner/work/pgwatch/pgwatch/cmd/pgwatch/main.go:100 +0x8eb
] runtime error: invalid memory address or nil pointer dereference (/home/runner/work/pgwatch/pgwatch/cmd/pgwatch/main.go:49 main.main.func1)
```

--- a/internal/sources/types.go
+++ b/internal/sources/types.go
@@ -2,6 +2,7 @@ package sources
 
 import (
 	"context"
+	"fmt"
 	"maps"
 	"reflect"
 	"slices"
@@ -64,6 +65,20 @@ type (
 
 	Sources []Source
 )
+
+func (srcs Sources) Validate() (Sources, error) {
+	names := map[string]any{}
+	for _, src := range srcs {
+		if _, ok := names[src.Name]; ok {
+			return nil, fmt.Errorf("duplicate source with name '%s' found", src.Name)
+		}
+		names[src.Name] = nil
+		if src.Kind == "" {
+			src.Kind = SourcePostgres
+		}
+	}
+	return srcs, nil
+}
 
 func (s *Source) GetDatabaseName() string {
 	if с, err := pgx.ParseConfig(s.ConnStr); err == nil {

--- a/internal/sources/yaml.go
+++ b/internal/sources/yaml.go
@@ -4,6 +4,7 @@ package sources
 
 import (
 	"context"
+	"fmt"
 	"io/fs"
 	"os"
 	"path/filepath"
@@ -61,6 +62,7 @@ func (fcr *fileSourcesReaderWriter) GetSources() (dbs Sources, err error) {
 	}
 	switch mode := fi.Mode(); {
 	case mode.IsDir():
+		var names []string
 		err = filepath.WalkDir(fcr.path, func(path string, d fs.DirEntry, err error) error {
 			if err != nil {
 				return err
@@ -70,10 +72,17 @@ func (fcr *fileSourcesReaderWriter) GetSources() (dbs Sources, err error) {
 				return nil
 			}
 			var mdbs Sources
-			if mdbs, err = fcr.getSources(path); err == nil {
-				dbs = append(dbs, mdbs...)
+			if mdbs, err = fcr.getSources(path); err != nil {
+				return err
 			}
-			return err
+			for _, mdb := range mdbs {
+				if slices.Contains(names, mdb.Name) {
+					return fmt.Errorf("duplicate source with name '%s' found in file %s", mdb.Name, path)
+				}
+				names = append(names, mdb.Name)
+			}
+			dbs = append(dbs, mdbs...)
+			return nil
 		})
 	case mode.IsRegular():
 		dbs, err = fcr.getSources(fcr.path)
@@ -93,7 +102,12 @@ func (fcr *fileSourcesReaderWriter) getSources(configFilePath string) (dbs Sourc
 	if err = yaml.Unmarshal(yamlFile, &c); err != nil {
 		return
 	}
+	var names []string
 	for _, v := range c {
+		if slices.Contains(names, v.Name) {
+			return nil, fmt.Errorf("duplicate source with name '%s' found in file %s", v.Name, configFilePath)
+		}
+		names = append(names, v.Name)
 		if v.Kind == "" {
 			v.Kind = SourcePostgres
 		}

--- a/internal/sources/yaml.go
+++ b/internal/sources/yaml.go
@@ -4,7 +4,6 @@ package sources
 
 import (
 	"context"
-	"fmt"
 	"io/fs"
 	"os"
 	"path/filepath"
@@ -62,7 +61,6 @@ func (fcr *fileSourcesReaderWriter) GetSources() (dbs Sources, err error) {
 	}
 	switch mode := fi.Mode(); {
 	case mode.IsDir():
-		var names []string
 		err = filepath.WalkDir(fcr.path, func(path string, d fs.DirEntry, err error) error {
 			if err != nil {
 				return err
@@ -72,17 +70,10 @@ func (fcr *fileSourcesReaderWriter) GetSources() (dbs Sources, err error) {
 				return nil
 			}
 			var mdbs Sources
-			if mdbs, err = fcr.getSources(path); err != nil {
-				return err
+			if mdbs, err = fcr.getSources(path); err == nil {
+				dbs = append(dbs, mdbs...)
 			}
-			for _, mdb := range mdbs {
-				if slices.Contains(names, mdb.Name) {
-					return fmt.Errorf("duplicate source with name '%s' found in file %s", mdb.Name, path)
-				}
-				names = append(names, mdb.Name)
-			}
-			dbs = append(dbs, mdbs...)
-			return nil
+			return err
 		})
 	case mode.IsRegular():
 		dbs, err = fcr.getSources(fcr.path)
@@ -90,7 +81,7 @@ func (fcr *fileSourcesReaderWriter) GetSources() (dbs Sources, err error) {
 	if err != nil {
 		return nil, err
 	}
-	return
+	return dbs.Validate()
 }
 
 func (fcr *fileSourcesReaderWriter) getSources(configFilePath string) (dbs Sources, err error) {
@@ -102,15 +93,7 @@ func (fcr *fileSourcesReaderWriter) getSources(configFilePath string) (dbs Sourc
 	if err = yaml.Unmarshal(yamlFile, &c); err != nil {
 		return
 	}
-	var names []string
 	for _, v := range c {
-		if slices.Contains(names, v.Name) {
-			return nil, fmt.Errorf("duplicate source with name '%s' found in file %s", v.Name, configFilePath)
-		}
-		names = append(names, v.Name)
-		if v.Kind == "" {
-			v.Kind = SourcePostgres
-		}
 		dbs = append(dbs, fcr.expandEnvVars(v))
 	}
 	return

--- a/internal/sources/yaml_test.go
+++ b/internal/sources/yaml_test.go
@@ -72,6 +72,51 @@ func TestYAMLGetMonitoredDatabases(t *testing.T) {
 		a.Error(err)
 		a.Nil(dbs)
 	})
+
+	t.Run("duplicate in single file", func(t *testing.T) {
+		tmpFile := filepath.Join(t.TempDir(), "duplicate.yaml")
+		yamlContent := `
+- name: test1
+  conn_str: postgresql://localhost/test1
+- name: test2
+  conn_str: postgresql://localhost/test2
+- name: test1
+  conn_str: postgresql://localhost/test1_duplicate
+`
+		err := os.WriteFile(tmpFile, []byte(yamlContent), 0644)
+		a.NoError(err)
+		yamlrw, err := sources.NewYAMLSourcesReaderWriter(ctx, tmpFile)
+		a.NoError(err)
+
+		dbs, err := yamlrw.GetSources()
+		a.Error(err)
+		a.Nil(dbs)
+	})
+
+	t.Run("duplicates across files", func(t *testing.T) {
+		tmpDir := t.TempDir()
+		yamlContent1 := `
+- name: test1
+  conn_str: postgresql://localhost/test1
+- name: test2
+  conn_str: postgresql://localhost/test2
+`
+		err := os.WriteFile(filepath.Join(tmpDir, "sources1.yaml"), []byte(yamlContent1), 0644)
+		a.NoError(err)
+
+		yamlContent2 := `
+- name: test1
+  conn_str: postgresql://localhost/test1_duplicate
+`
+		err = os.WriteFile(filepath.Join(tmpDir, "sources2.yaml"), []byte(yamlContent2), 0644)
+		a.NoError(err)
+		yamlrw, err := sources.NewYAMLSourcesReaderWriter(ctx, tmpDir)
+		a.NoError(err)
+
+		dbs, err := yamlrw.GetSources()
+		a.Error(err)
+		a.Nil(dbs)
+	})
 }
 
 func TestYAMLDeleteDatabase(t *testing.T) {

